### PR TITLE
Add support for quick unlock with TouchID on Macbook Pro

### PR DESCRIPTION
**Pull request by kolhagen**
## Description
Unlock your database using TouchID on supported Macbook Pro models.

## Motivation and context
With quick unlock one can simply unlock the database using TouchID without having to enter your password again while the application is still running. #209 

### Features
- TouchID is a option that can be activated on the unlock screen
- TouchID is only used for **temporary quick unlock**
  - Database has to be unlocked once after restart of the application in order to activate quick unlock using TouchID
- TouchID quick unlock **only replaces the password** part of the unlock process
  - With quick unlock activated the password will be encrypted (AES 256) and then kept in memory
  - The random encryption key (and IV) will be protected by the KeyChain using TouchID
- Next time you don't need to enter your password, just press "OK" to unlock your database with TouchID (however you will still need your additional database key, Yubikey, ...)

### Compilation
- Use `-DWITH_XC_TOUCHID=ON`
- Needs code-signature and entitlements in order to work (see issue reference)

### Possible future enhancements:
- Provide abstraction layer for other OS/fingerprint reader.
- Unlock with just putting on fingerprint w/o actively having to click unlock (KeyChain API may not allow this).
- Indicate if TouchID is currently active (i.e. no password is needed)
  - e.g. show special button on TouchBar

## How has this been tested?
- So far only manually on MBP with TouchID.

## Screenshots:
image

image

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

File: src/touchid/TouchID.mm [hunk 1]
```diff
@@ -0,0 +1,212 @@
+#define SECURITY_ATTRIBUTE_ACCOUNT CFSTR("KeepassXC TouchID Keys")
+#define TOUCH_ID_PROMPT_MESSAGE CFSTR("authenticate to access KeepassXC database!")
```
Comment (RealOrangeOne) [inline src/touchid/TouchID.mm]: From the screenshot, looks like the touchid UI automatically adds a _._ to the end of this message, making the exclamation mark seem out of place.

Comment (kolhagen) [reply src/touchid/TouchID.mm]: You're right, I'll remove it and replace the whole thing with QCoreApplication::translate(...).

**Comment by weslly** [comment]
Looks great! Unfortunately I don't have touch ID on my mac so this I'll be pretty hard to test :(

**Comment by kolhagen** [comment]
@weslly Does anyone of the development team have one to test with? Also, can you tell me why the build fails so I can make adjustments?

**Comment by weslly** [comment]
@kolhagen You can see the build log if you click in details and login as a guest at TeamCity

**Comment by kolhagen** [comment]
Did another rebase. Any updates on a possible integration?

**Comment by weslly** [comment]
@kolhagen I haven't found any easy way to simulate TouchID on macOS like you can do on iOS simulator, but it seems there's a workaround:

https://stackoverflow.com/questions/40414077/how-can-you-test-touch-id-for-macos

Can you add this policy so we can test it?

**Comment by droidmonkey** [comment]
I think we can merge this

**Comment by weslly** [comment]
Actually it already seems to hide the checkbox if I change the policy back to `DeviceOwnerAuthenticationWithBiometrics`

File: src/gui/DatabaseOpenWidget.cpp [hunk 2]
```diff
@@ -187,6 +199,24 @@ void DatabaseOpenWidget::openDatabase()
...
+        QHash<QString, QVariant> useTouchID = config()->get("UseTouchID").toHash();
+
+        // check if TouchID can & should be used to unlock the database next time
+        if (m_ui->checkTouchID->isChecked() && TouchID::getInstance().isAvailable()) {
+            // encrypt and store key blob
+            if (TouchID::getInstance().storeKey(m_filename, PasswordKey(m_ui->editPassword->text()).rawKey())) {
+                useTouchID.insert(m_filename, true);
+            }
+        } else {
+            // when TouchID not available or unchecked, reset for all databases
```
Comment (droidmonkey) [inline src/gui/DatabaseOpenWidget.cpp]: Technically this code only resets for the current database

File: src/touchid/TouchID.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+#include <QString>
+#include <QByteArray>
+#include <QSharedPointer>
+
+class TouchID
+{
+    public:
+        static TouchID& getInstance()
+        {
+            static TouchID instance; // Guaranteed to be destroyed.
```
Comment (droidmonkey) [inline src/touchid/TouchID.h]: I much prefer having this function body and global variable declared in TouchID.mm. Declaring it in the header may cause issues if we move this to an external library in the future.

**Comment by id-kemo** [comment]
In case you would need any help with testing the feature, I can provide some sort of feedback or help if needed as I have a Macbook Pro 15'' with TouchID (2017 model).

**Comment by kolhagen** [comment]
There's still an open [issue with the complexity](https://www.codefactor.io/repository/github/keepassxreboot/keepassxc/pull/1851) of the method `DatabaseOpenWidget::databaseKey()`.

Does it help splitting the method up (e.g. [like this](https://gist.github.com/kolhagen/606ff873e694bf1504d8e30781418d1a))?

**Comment by droidmonkey** [comment]
You can ignore that warning, I might remove codefactor as a check. You can't turn complexity to a warning....

**Comment by phoerious** [comment]
Hey, due to previous time constraints, I'm a little late to the review party. So first of all, thanks for your effort. I haven't really had a chance to look at the code yet, but I think at least the UI needs some tweaks. The most important issue is probably the alignment of the checkbox on the unlock screen. Please fix that before we merge. After this PR we should also have some discussion about a general redesign of that screen, since it starts becoming really crowded with all sorts of options.

**Comment by kolhagen** [comment]
Thanks for the feedback! Fixed the UI issue with checkboxes not being aligned and did another rebase.

**Comment by bhavers** [comment]
Great addition, thx! Happy to help testing, i have a MBP 15" 2017. Let me know when and what to do when you are ready.

**Comment by bhavers** [comment]
Hi, i just tried version 2.3.4 but couldn't find the touchid support. I thought there was nothing blocking the merger into a final version. Don't want to push anybody, but would be great if it could be in the next release.

**Comment by droidmonkey** [comment]
This is destined for 2.4, it will be merged very soon

**Comment by droidmonkey** [comment]
We will be doing a beta launch very soon. The last major feature was just integrated (group sharing).

**Comment by bhavers** [comment]
Hi, i tried 2.4 Beta 1 on my MBP 2017 (Mojave 10.14.3) but can't get TouchID to work, not sure if i am doing the right thing. Should i open a separate issue or can we use this thread?

Problem:
When i open KeepassXC i do see the TouchID option (see screenshot below).
I enter my password and open my database file. Than, when the session expires, i can't reopen it with TouchID. I have to provide the password again. 
If i just check TouchID and select OK it is unable to open the database.

image

image

image

**Comment by tijme** [comment]
> Hi, i tried 2.4 Beta 1 on my MBP 2017 (Mojave 10.14.3) but can't get TouchID to work, not sure if i am doing the right thing. Should i open a separate issue or can we use this thread?
> 
> Problem:
> When i open KeepassXC i do see the TouchID option (see screenshot below).
> I enter my password and open my database file. Than, when the session expires, i can't reopen it with TouchID. I have to provide the password again.
> If i just check TouchID and select OK it is unable to open the database.
> 
> image
> 
> image
> 
> image

Are you using the beta version (which is probably not signed)? I think @droidmonkey stated that TouchID does not work unless the binary is signed.

**Comment by droidmonkey** [comment]
I signed the beta version. Unfortunately I cannot test touch id since I only have a Mac mini. Please open a new issue, we may have broken it during one of our refactors.

**Comment by bhavers** [comment]
I opened a new bug [report for this](https://github.com/keepassxreboot/keepassxc/issues/2676).
Let me know how i can help you test if you don't have a TouchID device.

**Comment by bhavers** [comment]
Hi, just downloaded en tried beta 2. But still not able to log in with TouchID. Should i open a new bug report?

**Comment by droidmonkey** [comment]
@bhavers yes please, include screenshots of the unlock dialog

**Comment by bhavers** [comment]
Done, see [this bug report](https://github.com/keepassxreboot/keepassxc/issues/2720).

**Comment by mxk6n** [comment]
To further improve the utilization of TouchID on Macbook Pro and make the unlock feature more intuitive to use (see comments in #2720), I would propose implementing the following changes:

- Make enabling TouchID a **global setting** (not per database)
- Add an **indication that TouchID is available to unlock** (colored icon) on the unlock screen / greyed out when it is not. When it is disabled in the settings no icon is shown.
- Unlock with **TouchID is triggered** when it is activated/available and ...
  - ... User *clicks* on the colored icon
  - ... User tries to unlock database *without entering a password*
  - ... Window is being *brought to the foreground* by the user
  - ... *Tab is being changed* / brought to foreground (multi database usage)
  - I would suggest making the last two scenarios an optional setting
- TouchID is automatically **deactivated on 3 consecutively failed tries** (until user enters the password correctly again)
- **Raising the maximum time-out** of TouchID (currently the maximum value is 999 minutes)

How do you guys feel about this?

**Comment by droidmonkey** [comment]
Please make a new issue with your changes.

**Comment by alimsyed** [comment]
Hello, I have just built on my 2018 MBP with Touch ID. I have built with following options: ```
cmake -DCMAKE_OSX_ARCHITECTURES=x86_64 \
  -DCMAKE_BUILD_TYPE=Release \
  -DWITH_XC_ALL=ON -DWITH_XC_TOUCHID=ON
```

But I don't see the checkbox for "Touch ID" login. What am I doing wrong?

**Comment by augustgerro** [comment]
@mxk6n  Great feature. Thank you for implementing!

But I want to add that while using touch id is not very convenient. 

Here is my workflow:

1. when opening DB I see a window with a password to use Touch ID. And I should: 
2. click Return
3. the Touch ID window appears 
4. I put my finger on the sensor, logged in 

It would be more convenient to immediately request Touch ID, as is done in Safari. 

1. opened DB, a window with Touch ID 
2. I put my finger on the sensor, logged in

**Comment by anonymous133** [comment]
I have my new Macbook since 4 months (first time with fingerprint sensor) and I just found out how it works with keepass after investing 10 minutes into finding how to activate it just to find out by myself that I have to click OK with an empty password. I think the OK button should have the text "Unlock with fingerprint" if no password is entered and the option is active. Or like @augustgerro said to just bring it up whenever the db is closed.

**Pull request merged**